File: agent-connect-android/app/src/main/java/sg/searchhouse/agentconnect/model/api/agent/AgentPO.kt
package sg.searchhouse.agentconnect.model.api.agent

import android.content.Context
import android.text.TextUtils
import com.google.gson.annotations.SerializedName
import sg.searchhouse.agentconnect.R
import sg.searchhouse.agentconnect.util.StringUtil
import java.io.Serializable

data class AgentPO constructor(
    @SerializedName("id")
    val id: String,
    @SerializedName("userId")
    val userId: Int,
    @SerializedName("encryptedUserId")
    val encryptedUserId: String,
    @SerializedName("name")
    val name: String,
    @SerializedName("ceaRegNo")
    val ceaRegNo: String,
    @SerializedName("mobile")
    val mobile: String,
    @SerializedName("agency")
    val agency: String,
    @SerializedName("agencyId")
    val agencyId: Int? = null,
    @SerializedName("agencyName")
    val agencyName: String,
    @SerializedName("licenseNumber")
    val licenseNumber: String,
    @SerializedName("email")
    val email: String,
    @SerializedName("photo")
    val photo: String,
    @SerializedName("agencyLogo")
    val agencyLogo: String,
    @SerializedName("subscription")
    val subscription: String,
    @SerializedName("subscriptionType")
    val subscriptionType: String,
    @SerializedName("numOfListings")
    val numOfListings: Int,
    @SerializedName("agentCvPO")
    val agentCvPO: AgentCvPO?,
    @SerializedName("agentCvUrl")
    val agentCvUrl: String,
    @SerializedName("agencyPartOfSRXConsortium")
    val agencyPartOfSRXConsortium: Boolean,
    @SerializedName("transactionSummary")
    val transactionSummary: TransactionSummary?,
    @SerializedName("listingSummary")
    val listingSummary: ListingSummary?,

    var isSelected: Boolean = false
) : Serializable {

    data class TransactionSummary(
        @SerializedName("rentHdbTotal")
        val rentHdbTotal: Int,
        @SerializedName("rentHighest")
        val rentHighest: Int,
        @SerializedName("rentPrivateTotal")
        val rentPrivateTotal: Int,
        @SerializedName("rentTotal")
        val rentTotal: Int,
        @SerializedName("saleHdbTotal")
        val saleHdbTotal: Int,
        @SerializedName("saleHighest")
        val saleHighest: Int,
        @SerializedName("salePrivateTotal")
        val salePrivateTotal: Int,
        @SerializedName("saleTotal")
        val saleTotal: Int
    ) : Serializable

    data class ListingSummary(
        @SerializedName("rentTotal")
        val rentTotal: Int,
        @SerializedName("saleTotal")
        val saleTotal: Int
    ) : Serializable

    fun isExpertAnalyzer(): Boolean {
        return (subscription == "HRD" || subscription == "Expert")
    }

    fun getCeaNoAndLicenseNumber(): String {
        return StringBuilder().append(ceaRegNo).append(" / ").append(licenseNumber).toString()
    }

    fun getActiveListings(): String {
        var activeListings = ""
        if (numOfListings > 0) {
            activeListings = when (numOfListings) {
                1 -> {
                    "$numOfListings Active Listing"
                }
                else -> {
                    "$numOfListings Active Listings"
                }
            }
        }
        return activeListings
    }

    // Handle case when CEA empty
    fun ceaRegNumberFormatted(context: Context): String {
        return if (TextUtils.isEmpty(ceaRegNo)) {
            context.getString(R.string.label_cea_not_available)
        } else {
            ceaRegNo
        }
    }

    fun getCeaNoWithCEAInitial(context: Context): String {
        return if (TextUtils.isEmpty(ceaRegNo)) {
            context.getString(R.string.label_cea_not_available)
        } else {
            "CEA : $ceaRegNo"
        }
    }

    // Sanitized Mobile Number
    fun getSanitizedMobileNumber(): String? {
        return StringUtil.getSanitizedMobileNumber(mobile)
    }

    // Sanitized Mobile Number with SG country code
    fun getSanitizedMobileNumberWithSgCountryCode(): String? {
        return StringUtil.getSanitizedMobileNumberWithSgCountryCode(mobile)
    }

    fun getMobileNumberForDisplay(): String {
        getSanitizedMobileNumber()?.let {
            return it.substring(0, 4).plus(" XXXX")
        }
        return ""
    }
}

File: srxproperty/src/dataObject/AgentPO.js
import { CommonUtil, ObjectUtil } from "../utils";
import { AgentCvPO } from "./AgentCvPO";
import { ListingSummaryPO } from "./ListingSummaryPO";
import { TransactionSummaryPO } from "./TransactionSummaryPO";

class AgentPO {
  constructor(data) {
    if (data) {
      const {
        agency,
        agencyName,
        agencyLogo,
        agentCvPO,
        agentCvUrl,
        ceaRegNo,
        contactNumber, //from contactInformation
        email,
        emailAddress, //from contactInformation
        encryptedUserId,
        eyecatcherMsg,
        id,
        licenseNumber,
        listingSummary,
        mobile,
        name,
        numOfListings,
        photo,
        subscription,
        trackingId,
        transactionSummary,
        userId
      } = data;
      this.agency = agency;
      this.agencyName = agencyName;
      this.agencyLogo = agencyLogo;

      this.agentCvPO = null;
      if (!ObjectUtil.isEmpty(agentCvPO)) {
        this.agentCvPO = new AgentCvPO(agentCvPO);
      }

      this.agentCvUrl = agentCvUrl;
      this.ceaRegNo = ceaRegNo;

      if (mobile) this.mobile = mobile;
      else if (contactNumber) this.mobile = contactNumber;

      if (email) this.email = email;
      else if (emailAddress) this.email = emailAddress;

      this.encryptedUserId = encryptedUserId;
      this.eyecatcherMsg = eyecatcherMsg;
      this.id = id;
      this.mobile = mobile;

      if (id) this.id = id;
      else if (userId) this.id = userId;

      this.listingSummary = new ListingSummaryPO(listingSummary);
      this.licenseNumber = licenseNumber;
      this.displayFullMobile = false; //self define variable
      this.name = name;
      this.numOfListings = numOfListings;
      this.photo = photo;
      this.subscription = subscription;
      this.trackingId = trackingId;

      this.transactionSummary = new TransactionSummaryPO(transactionSummary);

      this.userId = userId;
    }
  }

  getAgentId = () => {
    return this.userId;
  };

  getAgentCvPO = () => {
    const { agentCvPO } = this;
    return agentCvPO;
  };

  getMobileNumber = () => {
    if (this.mobile) {
      var mobileString = "";
      if (typeof this.mobile === "string") {
        mobileString = this.mobile.replace(/[\(\)]+/g, "");
      } else if (typeof this.mobile === "number") {
        mobileString = this.mobile.toString();
      }
    }

    return mobileString;
  };

  getMobileNumberForDisplay = () => {
    var mobileString = this.getMobileNumber();

    if (this.displayFullMobile) {
      return mobileString;
    } else {
      if (mobileString.length > 4) {
        return mobileString.substring(0, 4) + " XXXX";
      }
    }

    return mobileString;
  };

  isExpert = () => {
    return (
      this.subscription &&
      (this.subscription === "HRD" || this.subscription === "Expert")
    );
  };

  getAgentPhoto = () => {
    if (!ObjectUtil.isEmpty(this.photo)) {
      return CommonUtil.handleImageUrl(this.photo);
    } else if (!ObjectUtil.isEmpty(this.id) || this.id > 0) {
      return CommonUtil.handleImageUrl(
        "/mobile/iphone/photo.mobile3?type=A&userId=" +
          this.id +
          "&userid=" +
          this.id
      );
    }
    return "";
  };

  getAgencyLogoURL = () => {
    if (!ObjectUtil.isEmpty(this.agencyLogo)) {
      return CommonUtil.handleImageUrl(this.agencyLogo);
    }
    return "";
  };

  getExpertAnalyzerAgent = () => {
    let expertSubscriptionString = "";
    if (this.isExpert()) {
      expertSubscriptionString = this.subscription + " Analyzer";
    }
    return expertSubscriptionString;
  };

  getActiveListings = () => {
    let activeListings = "";
    if (this.numOfListings && this.numOfListings > 0) {
      if (this.numOfListings === 1) {
        activeListings = this.numOfListings + " Active Listing";
      } else {
        activeListings = this.numOfListings + " Active Listings";
      }
    }
    return activeListings;
  };
}

export { AgentPO };
